core/sql/common/CharType.cpp

```diff
@@ -967,6 +972,28 @@ SQLVarChar::SQLVarChar(const CharLenInfo & maxLenInfo,
 {}
 
 // -----------------------------------------------------------------------
+// Equality comparison
+// -----------------------------------------------------------------------
+NABoolean SQLVarChar::operator==(const NAType& other) const
+{
+  NABoolean compResult = FALSE; // false, failed. true, passed.
+
+  // if this or other was originally a hive string type, then
+  // do not compare lengths.
+  if ((wasHiveString()) ||
+      ((other.getTypeName() == LiteralVARCHAR) &&
+       (((SQLVarChar&)other).wasHiveString())))
```

Comment:
literalVARCHAR is used for all sql varchar datatypes.
The bytesPerChar and character set parameters to the char
classes defined in common/CharType.h determine what
kind of char datatype it is.